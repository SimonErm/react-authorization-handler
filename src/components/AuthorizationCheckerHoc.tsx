import React, { ReactNode } from "react";
import { useAuthority } from "../hooks/AuthorityHook";

export function AuthorizationChecker<P extends string>({
  allowedRoles,
  children,
  fallback,
  requiredRole,
  requiredPermission
}: {
  allowedRoles?: string[];
  fallback?: React.ReactElement;
  requiredPermission?: P;
  requiredRole?: string;
  children: ReactNode;
}) {
  const { hasPermission, isRequiredRole, isAllowedRole } = useAuthority();
  const isGranted =
    (requiredRole === undefined &&
      requiredPermission === undefined &&
      allowedRoles === undefined) ||
    (requiredRole !== undefined && isRequiredRole(requiredRole)) ||
    (requiredPermission !== undefined && hasPermission(requiredPermission)) ||
    (allowedRoles !== undefined && isAllowedRole(allowedRoles));

  if (isGranted) {
    return <>{children}</>;
  }
  if (fallback) {
    return fallback;
  }
  return null;
}
