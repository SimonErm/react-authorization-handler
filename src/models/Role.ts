export interface Role<T extends string> {
  name: string;
  permissions?: T[];
}
