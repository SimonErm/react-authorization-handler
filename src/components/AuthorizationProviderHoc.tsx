import { AuthentificationContext } from "../context/AuthentificationContext";
import { FC, useState } from "react";
import { Role } from "../models/Role";
import * as React from "react";
export const AuthorizationProvider: FC<{ initialRole?: Role<string> }> = ({
  children,
  // eslint-disable-next-line react/prop-types
  initialRole = { name: "Guest" }
}) => {
  const [role, setRole] = useState(initialRole);

  return (
    <AuthentificationContext.Provider value={{ role, setRole }}>
      {children}
    </AuthentificationContext.Provider>
  );
};
