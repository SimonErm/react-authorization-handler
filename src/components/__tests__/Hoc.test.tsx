import React from "react";
import renderer from "react-test-renderer";
import { AuthorizationProvider } from "../AuthorizationProviderHoc";
import { AuthorizationChecker } from "../AuthorizationCheckerHoc";
import { Role } from "../../models/Role";
type TestPermissions = "askForTest" | "doThis" | "doThat";
const defaultRole: Role<TestPermissions> = {
  name: "Guest",
  permissions: ["askForTest"]
};
describe("Test ProviderHoc", () => {
  test("Target should be null", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker requiredPermission="test">
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Target should be Fallback", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker
          requiredPermission="test"
          fallback={
            <div>
              <h1>Fallback</h1>
            </div>
          }
        >
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("[requiredPermission]Target should be rendered by permission", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker<TestPermissions> requiredPermission="askForTest">
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("[requiredRole]Target should be rendered by role", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker requiredRole="Guest">
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("[requiredRole]Target should not be rendered by role", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker<TestPermissions> requiredRole="Admin">
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("[allowedRoles]Target should be rendered by role", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker allowedRoles={["Guest", "Another"]}>
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("[allowedRoles]Target should not be rendered by role", () => {
    const component = renderer.create(
      <AuthorizationProvider initialRole={defaultRole}>
        <AuthorizationChecker allowedRoles={["Admin", "Another"]}>
          <div>
            <h1>isGranted</h1>
          </div>
        </AuthorizationChecker>
      </AuthorizationProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
