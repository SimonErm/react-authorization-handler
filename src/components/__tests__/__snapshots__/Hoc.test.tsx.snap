// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ProviderHoc [allowedRoles]Target should be rendered by role 1`] = `
<div>
  <h1>
    isGranted
  </h1>
</div>
`;

exports[`Test ProviderHoc [allowedRoles]Target should not be rendered by role 1`] = `null`;

exports[`Test ProviderHoc [requiredPermission]Target should be rendered by permission 1`] = `
<div>
  <h1>
    isGranted
  </h1>
</div>
`;

exports[`Test ProviderHoc [requiredRole]Target should be rendered by role 1`] = `
<div>
  <h1>
    isGranted
  </h1>
</div>
`;

exports[`Test ProviderHoc [requiredRole]Target should not be rendered by role 1`] = `null`;

exports[`Test ProviderHoc Target should be Fallback 1`] = `
<div>
  <h1>
    Fallback
  </h1>
</div>
`;

exports[`Test ProviderHoc Target should be null 1`] = `null`;
